src/ua_parser/caching.py

```diff
@@ -0,0 +1,128 @@
+    """
+
+    @abc.abstractmethod
+    def __setitem__(self, key: str, value: PartialParseResult):
+        """Adds or replace ``value`` to the cache at key ``key``."""
+        ...
+
+    @abc.abstractmethod
+    def __getitem__(self, key: str) -> Optional[PartialParseResult]:
+        """Returns a partial result for ``key`` if there is any."""
+        ...
+
+
+class Clearing(Cache):
+    """A clearing cache, if the cache is full, just remove all the entries
+    and re-fill from scratch.
+
+    This can also be used as a cheap permanent cache by setting the
+    ``maxsize`` to infinity (or at least some very large value),
+    however this is probably a bad idea as it *will* lead to an
+    ever-growing memory allocation, until every possible user agent
+    string has been seen.
+    """
+
+    def __init__(self, maxsize: int):
+        self.maxsize = maxsize
+        self.cache: Dict[str, PartialParseResult] = {}
+
+    def __getitem__(self, key: str) -> Optional[PartialParseResult]:
+        return self.cache.get(key)
+
+    def __setitem__(self, key: str, value: PartialParseResult):
+        if key not in self.cache and len(self.cache) >= self.maxsize:
+            self.cache.clear()
+
+        self.cache[key] = value
+
+
+class LRU(Cache):
+    """Cache following a least-recently used replacement policy: when
+    there is no more room in the cache, whichever entry was last seen
+    the least recently is removed.
+
+    Note that the cache size is adjusted after inserting the new
+    entry, so the cache will temporarily contain ``maxsize + 1``
+    items.
+    """
+
+    def __init__(self, maxsize: int):
+        self.maxsize = maxsize
+        self.cache: OrderedDict[str, PartialParseResult] = OrderedDict()
+
+    def __getitem__(self, key: str) -> Optional[PartialParseResult]:
+        e = self.cache.get(key)
+        if e:
+            self.cache.move_to_end(key)
+        return e
+
+    def __setitem__(self, key: str, value: PartialParseResult):
+        self.cache[key] = value
+        self.cache.move_to_end(key)
+        while len(self.cache) > self.maxsize:
+            self.cache.popitem(last=False)
+
+
+Lock = Callable[[], ContextManager]
+
+
+class CachingParser(Parser):
+    """A wrapping parser which takes an underlying concrete :class:`Cache`
+    for the actual caching and cache strategy.
+
+    The :class:`CachingParser` only interacts with the :class:`Cache`
+    and delegates to the wrapped parser in case of lookup failure.
+
+    :class:`CachingParser` will set entries back in the cache when
+    filling them up, it does not update results in place (and can't
+    really, they're immutable).
+    """
+
+    def __init__(self, parser: Parser, cache: Cache, lock: Lock = threading.Lock):
+        self.parser: Parser = parser
+        self.cache: Cache = cache
+        self.lock: ContextManager = lock()
+
+    def __call__(self, ua: str, domains: Domain, /) -> PartialParseResult:
+        with self.lock:
```

Comment:
This is just locking the `__getitem__` and `__setitem__`, why not just have a locking `Cache` instead?